### PR TITLE
[ADD] beesdoo_worker_status: config to disable penalty

### PR DESCRIPTION
**Pull request by victor-champonnois**
## Odoo task (if applicable)
Task https://gestion.coopiteasy.be/web#id=9394&action=479&model=project.task&view_type=form&menu_id=536
Spec https://docs.google.com/document/d/1da5VDkctngDbjrZ6WDsuj-x5_kaTKf0z68Ac9Y_YuB0/edit#heading=h.jsre61isklc0

## Checklist before approval

- [ ] Tests are present (or not needed).
- [ ] Credits/copyright have been changed correctly.
- [ ] Change log snippet is present.
- [ ] (If a new module) Moving this to OCA has been considered.

### CODE REVIEW

File: beesdoo_worker_status/models/cooperative_status.py [hunk 1]
```diff
@@ -350,16 +350,19 @@ def _get_irregular_worker_domain(self, today):
...
+            if self.sr > 0:
+                if self.sr >= 1:
+                    self.is_penalised_irregular = False
+                self.sr -= 1
+            elif self.alert_start_time or self.is_penalised_irregular:
+                self.sr -= 1
+            else:
+                self.sr -= 2
+                # Until sr is 1 or above, don't do another penalty.
+                self.is_penalised_irregular = True
```
Comment (robinkeunen) [inline beesdoo_worker_status/models/cooperative_status.py]: This seems more readable to me :

```suggestion
        irregular_penalty =  self.env["ir.config_parameter"].sudo().get_param("irregular_penalty")
        if irregular_penalty:
            if self.sr >= 1:
                self.sr -= 1
                self.is_penalised_irregular = False    
            elif self.sr == 0:
                self.sr -= 1
            elif self.alert_start_time or self.is_penalised_irregular:
                self.sr -= 1
            else:
                self.sr -= 2
                # Until sr is 1 or above, don't do another penalty.
                self.is_penalised_irregular = True
```

File: beesdoo_worker_status/models/res_config_settings.py [hunk 2]
```diff
@@ -0,0 +1,11 @@
+from odoo import fields, models
+
+
+class ResConfigSettings(models.TransientModel):
+    _inherit = "res.config.settings"
+
+    irregular_penalty = fields.Boolean(
+        string="Penalty for irregular with negative counter",
```
Comment (robinkeunen) [inline beesdoo_worker_status/models/res_config_settings.py]: Penalty for irregular worker with negative counter

**Review by robinkeunen — CHANGES_REQUESTED** [review]
readability comment otherwise LGTM, thx for the tests.

File: beesdoo_worker_status/models/res_config_settings.py (hunk 2 above)
Comment (robinkeunen) [inline beesdoo_worker_status/models/res_config_settings.py]: You're missing the help message : "Si cette option est sélectionnée, le compteur du volant passera de 0 à -2, sauf s’il a été en alerte juste avant, auquel cas il passera de 0 à -1"

File: beesdoo_worker_status/models/cooperative_status.py (hunk 1 above)
Comment (robinkeunen) [reply beesdoo_worker_status/models/cooperative_status.py]: Or 

```py
        irregular_penalty =  self.env["ir.config_parameter"].sudo().get_param("irregular_penalty")
        if irregular_penalty:
            if self.sr < 0 and not (self.alert_start_time or self.is_penalised_irregular):
                self.sr -= 2
                # Until sr is 1 or above, don't do another penalty.
                self.is_penalised_irregular = True
            else:
                if self.sr >= 1:
                    self.is_penalised_irregular = False
                self.sr -= 1
```

File: beesdoo_worker_status/models/res_config_settings.py (hunk 2 above)
Comment (victor-champonnois) [reply beesdoo_worker_status/models/res_config_settings.py]: fixed

Comment (victor-champonnois) [reply beesdoo_worker_status/models/res_config_settings.py]: fixed (translated in english)

File: beesdoo_worker_status/models/cooperative_status.py (hunk 1 above)
Comment (victor-champonnois) [reply beesdoo_worker_status/models/cooperative_status.py]: Fixed (with first version, this method is complex, I prefer avoiding risks to change it too much)

**Comment by victor-champonnois** [comment]
@robinkeunen Thanks for the review, all is fixed !

**Review by robinkeunen — APPROVED** [review]

File: beesdoo_worker_status/models/res_config_settings.py [hunk 3]
```diff
@@ -0,0 +1,14 @@
+from odoo import fields, models
+
+
+class ResConfigSettings(models.TransientModel):
+    _inherit = "res.config.settings"
+
+    irregular_penalty = fields.Boolean(
+        string="Penalty for irregular worker with negative counter",
+        help="""When selected, the irregar worker's counter will decrease
+        by two when zero or belo, unless they were in alert
```
Comment (carmenbianca) [inline beesdoo_worker_status/models/res_config_settings.py]: s/belo/below

File: beesdoo_worker_status/tests/test_beesdoo_shift.py [hunk 4]
```diff
@@ -42,6 +42,9 @@ def setUp(self):
     def test_shift_counters(self):
         "Test shift counters calculation and cooperative status update"
 
+        config = self.env["ir.config_parameter"].sudo()
+        config.set_param("irregular_penalty", True)
+
```
Comment (carmenbianca) [inline beesdoo_worker_status/tests/test_beesdoo_shift.py]: Let's put this in the setUp function and get rid of it here and elsewhere. We simply set it to False for exceptional cases.

**Comment by carmenbianca** [comment]
I tried everything under the sun to get the `irregular_penalty` to correctly save in the res.config.settings UI. Turns out: because it was set as `default=True` in the res.config.settings TransientModel, the _transient_ value of `irregular_penalty` defaulted to True, every time the settings UI is opened.

I fixed this by removing `default=True` and creating a `data/beesdoo_worker_status_data.xml` file where I set a default (noupdate=1) value instead.

While trying everything under the sun, I made lots of other changes. I decided to keep some of them.

**Comment by victor-champonnois** [comment]
@carmenbianca nice, thanks for the changes :)

File: beesdoo_worker_status/tests/test_beesdoo_shift.py (hunk 4 above)
Comment (victor-champonnois) [reply beesdoo_worker_status/tests/test_beesdoo_shift.py]: done

**Comment by victor-champonnois** [comment]
@carmenbianca is everything OK for you ?

**Review by carmenbianca — APPROVED** [review]
@victor-champonnois yes LGTM thank you victor

**Pull request merged**